internal/aws/xray/telemetry/recorder.go

```diff
@@ -0,0 +1,327 @@
+	Start()
+	// Stop send loop.
+	Stop()
+	// RecordSegmentsReceived adds the count to the record for the current period.
+	RecordSegmentsReceived(count int)
+	// RecordSegmentsSent adds the count to the record for the current period.
+	RecordSegmentsSent(count int)
+	// RecordSegmentsSpillover adds the count to the record for the current period.
+	RecordSegmentsSpillover(count int)
+	// RecordSegmentsRejected adds the count to the record for the current period.
+	RecordSegmentsRejected(count int)
+	// RecordConnectionError categorizes the error and increments the count by one for the current period.
+	RecordConnectionError(err error)
+}
+
+type telemetryRecorder struct {
+	resourceARN string
+	instanceID  string
+	hostname    string
+
+	// client is used to send the records.
+	client awsxray.XRayClient
+	// record is the pointer to the count metrics for the current period.
+	record *xray.TelemetryRecord
+	// mu is the lock used when updating the record. Primarily exists to
+	// prevent writes while swapping the record out in cutoff.
+	mu sync.Mutex
+
+	// queue is used to keep records that failed to send for retry during
+	// the next period.
+	queue chan *xray.TelemetryRecord
+	// done is the channel used to stop the loop.
+	done chan struct{}
+	// interval is the amount of time between flushes.
+	interval time.Duration
+
+	// recordUpdated is set to true when any count is updated. Indicates
+	// that telemetry data is available.
+	recordUpdated *atomic.Bool
+
+	startOnce sync.Once
+	stopOnce  sync.Once
+}
+
+func newTelemetryRecorder(
+	client awsxray.XRayClient,
+	sess *session.Session,
+	cfg *Config,
+	settings *awsutil.AWSSessionSettings,
+) Recorder {
+	recorder := &telemetryRecorder{
+		client:        client,
+		record:        newTelemetryRecord(),
+		recordUpdated: &atomic.Bool{},
+		interval:      defaultInterval,
+		done:          make(chan struct{}),
+		queue:         make(chan *xray.TelemetryRecord, queueSize),
+	}
+	if cfg != nil && cfg.IncludeMetadata {
+		if settings != nil {
+			recorder.resourceARN = settings.ResourceARN
+		}
+		recorder.hostname = os.Getenv(envAWSHostname)
+		recorder.instanceID = os.Getenv(envAWSInstanceID)
+
+		if settings == nil || !settings.LocalMode {
+			metadataClient := ec2metadata.New(sess)
+			if recorder.hostname == "" {
+				if hostname, err := metadataClient.GetMetadata("hostname"); err == nil {
+					recorder.hostname = hostname
+				}
+			}
+			if recorder.instanceID == "" {
+				if instanceID, err := metadataClient.GetMetadata("instance-id"); err == nil {
+					recorder.instanceID = instanceID
+				}
+			}
+		}
+	}
+	return recorder
+}
+
+func newTelemetryRecord() *xray.TelemetryRecord {
+	return &xray.TelemetryRecord{
+		SegmentsReceivedCount:  aws.Int64(0),
+		SegmentsRejectedCount:  aws.Int64(0),
+		SegmentsSentCount:      aws.Int64(0),
+		SegmentsSpilloverCount: aws.Int64(0),
+		BackendConnectionErrors: &xray.BackendConnectionErrors{
+			HTTPCode4XXCount:       aws.Int64(0),
```

Comment:
details metrics such as 403 and 429 are useful.

Reply:
Unfortunately, the `PutTelemetryRecord` (https://docs.aws.amazon.com/xray/latest/api/API_PutTelemetryRecords.html) doesn't have specific status code metrics.